freqtrade/freqai/freqai_interface.py

```diff
@@ -0,0 +1,155 @@
+        """
+        Entry point to the FreqaiModel, it will train a new model if 
+        necesssary before making the prediction.
+        The backtesting and training paradigm is a sliding training window
+        with a following backtest window. Both windows slide according to the
+        length of the backtest window. This function is not intended to be 
+        overridden by children of IFreqaiModel, but technically, it can be 
+        if the user wishes to make deeper changes to the sliding window
+        logic.
+        :params:
+        :dataframe: Full dataframe coming from strategy - it contains entire
+        backtesting timerange + additional historical data necessary to train 
+        the model.
+        :metadata: pair metadataa coming from strategy. 
+        """
+        self.pair = metadata['pair']
+        self.dh = DataHandler(self.config, dataframe, self.data)
+
+        print('going to train',len(self.dh.training_timeranges),
+            'timeranges:',self.dh.training_timeranges)
+
+        # Loop enforcing the sliding window training/backtesting paragigm
+        # tr_train is the training time range e.g. 1 historical month
+        # tr_backtest is the backtesting time range e.g. the week directly 
+        # following tr_train. Both of these windows slide through the 
+        # entire backtest
+        for tr_train, tr_backtest in zip(self.dh.training_timeranges,
+                                         self.dh.backtesting_timeranges):
+            gc.collect()
+            #self.config['timerange'] = tr_train
+            self.dh.data = {} # clean the pair specific data between models
+            self.freqai_info['training_timerange'] = tr_train
+            dataframe_train = self.dh.slice_dataframe(tr_train, dataframe)
+            dataframe_backtest = self.dh.slice_dataframe(tr_backtest, dataframe)
+            print("training",self.pair,"for",tr_train)
+            self.dh.model_path = self.full_path+"/"+ 'sub-train'+'-'+str(tr_train)+'/'
+            if not self.model_exists(self.pair, training_timerange=tr_train):
+                self.model = self.train(dataframe_train, metadata)
+                self.dh.save_data(self.model)
+            else:
+                self.model = self.dh.load_data(self.dh.model_path)
+
+            preds, do_preds = self.predict(dataframe_backtest)
+
+            self.dh.append_predictions(preds,do_preds,len(dataframe_backtest))
+        
+        self.dh.fill_predictions(len(dataframe))
+
+        return self.dh.predictions, self.dh.do_predict, self.dh.target_mean, self.dh.target_std
+
+    def make_labels(self, dataframe: DataFrame) -> DataFrame:
+        """
+        User defines the labels here (target values).
+        :params:
+        :dataframe: the full dataframe for the present training period
+        """
+
+        return dataframe
+
+    def train(self, unfiltered_dataframe: DataFrame, metadata: dict) -> Any:
```

Comment:
You should make this an @abstractmethod
that will force the inheriting class to override this - and i don't think calling using Any as model will work down the strategy.

Reply:
the return typing is fine i guess (maybe some scikit model supertype could work ...?) - it's about the `return Any` statement itself, really.

If it's an abstract method, you can leave the body empty without problem - so i'd do that.

Reply:
Thanks, fixed.